### ISSUE TITLE
System tray icon isn't placed correctly on some Linux desktop environments

### ISSUE THREAD

**Issue by nathan-osman** [nitroshare/nitroshare-desktop#1]
When running the application on Ubuntu 14.10, the system tray icon is not displayed in the expected location (with other application indicators) - instead, it is displayed in the upper-left corner of the screen. This is a [known bug](https://bugs.launchpad.net/sni-qt/+bug/1209106) and unfortunately, it will not be fixed for Ubuntu 14.04 & 14.10:

> "It is impossible to modify it to support Qt < 5.4, and we are not going to backport Qt 5.4 to Utopic. Sorry."
> — [Dmitry Shachnev](https://bugs.launchpad.net/sni-qt/+bug/1209106/comments/14)

Therefore, the only way to correct the behavior within these Ubuntu releases is to manually create an application indicator with D-Bus.

**Comment by nathan-osman** [comment]
I looked into what would be involved with accessing D-Bus directly and it wasn't going to be worth the effort. So instead, we will need to avoid displaying an indicator when running under Unity. I'm still considering what options are available as an alternative.

**Comment by mateosalta** [comment]
Maybe a quick list for the icon?

**Comment by nathan-osman** [comment]
We have a window that displays transfers that are in progress now. It's even a permanent window (it exists for as long as NitroShare is running). Perhaps that can be the way users interact with the application on Linux. (Windows users can still use the tray.) I think we might also have this problem on OS X. (I can't quite remember.)

**Comment by mateosalta** [comment]
I remember, that on the last one, on the mac, you could have both a tray icon and a dock icon, and It had a setting of turning off the dock icon. And had programing for right clicking on the icon.

https://c1.staticflickr.com/9/8177/8027230856_4034833c7d_m.jpg

So that mitght help for the convient acess that the tray provided, to do something similar for unity/docks

**Comment by nathan-osman** [comment]
Ah, okay. In that case, would it make sense _just_ to have the dock icon? Clicking on that would display the transfer window. It would also work the same way on Linux - since we only have the launcher icon in Unity. Windows may be the only platform where a tray icon makes sense. (And I'm open to other options there as well.)

**Comment by mateosalta** [comment]
I think that sounds reasonable,  maybe the main window could have some of the same stuff in the menus, hidden away just in case.

**Comment by nathan-osman** [comment]
Interestingly, I found [this forum post](http://ubuntuforums.org/showthread.php?t=2179045) which describes a method for displaying an application indicator within Unity with little effort. The best part is that it worked perfectly under Unity when I tested it. I will give this some more thought.

**Comment by nathan-osman** [comment]
I'm happy to report that this has now been fixed. I created an abstract base class named `Icon` that represents a dock/panel/tray icon of some sort. It requires each derived class to implement three methods:

- `void addAction(...)` - adds an action to the icon's menu
- `void addSeparator()` - adds a menu separator to the icon's menu
- `void showMessage()` - displays a notification message

Whenever the application is _not_ running under Unity, the `TrayIcon` class is used. It basically wraps `QSystemTrayIcon`. However, when running under Unity, the `IndicatorIcon` class is used instead. It uses the C functions from GTK/libappindicator to create an application indicator.
